rasa_nlu/featurizers/ngram_featurizer.py

```diff
@@ -0,0 +1,271 @@
+            candidates = []
+            features[n] = []
+            counters[n] = Counter()
+
+            # generate all possible n length ngrams
+            for text in list_of_strings:
+                text = text.replace(',', ' ')
+                for word in text.lower().split(' '):
+                    cands = [word[i:i + n] for i in xrange(len(word) - n)]
+                    for cand in cands:
+                        counters[n][cand] += 1
+                        if cand not in candidates:
+                            candidates.append(cand)
+
+            # iterate over these candidates picking only the applicable ones
+            for can in candidates:
+                if counters[n][can] > self.n_gram_min_occurrences:
+                    features[n].append(can)
+                    begin = can[:-1]
+                    end = can[1:]
+                    if n > self.n_gram_min_length:
+                        if counters[n - 1][begin] == counters[n][can] and begin in features[n - 1]:
+                            features[n - 1].remove(begin)
+                        if counters[n - 1][end] == counters[n][can] and end in features[n - 1]:
+                            features[n - 1].remove(end)
+
+        return [item for sublist in features.values() for item in sublist]
+
+    def _create_bow_vecs(self, intent_features, sentences, spacy_nlp, max_ngrams=None):
+        """Given a set of sentences, returns a feature vector for each sentence.
+
+        The first $k$ elements are from the `intent_features`,
+        the rest are {1,0} elements denoting whether an ngram is in sentence."""
+
+        import numpy as np
+
+        ngrams_to_use = self._ngrams_to_use(max_ngrams)
+        if not ngrams_to_use:
+            return intent_features
+        extras = np.array(self._ngrams_in_sentences(sentences, spacy_nlp, ngrams=ngrams_to_use))
+        total = np.hstack((intent_features, extras))
+        return total
+
+    def _cross_validation(self, sentences, labels, intent_features, spacy_nlp, max_ngrams):
+        """choose the best number of ngrams to include in bow.
+
+        Given an intent classification problem and a set of ordered ngrams (ordered in terms
+        of importance by pick_applicable_ngrams) we choose the best number of ngrams to include
+        in our bow vecs by cross validation."""
+
+        from sklearn import preprocessing
+        from sklearn.linear_model import LogisticRegression
+        from sklearn.model_selection import cross_val_score
+        import numpy as np
+
+        clf2 = LogisticRegression()
+        intent_encoder = preprocessing.LabelEncoder()
+        intent_encoder.fit(labels)
+        y = intent_encoder.transform(labels)
+        cv_splits = min(10, np.min(np.bincount(y)))
+        if cv_splits >= 3:
+            logging.debug("Started ngram cross-validation to find best number of ngrams to use...")
+            triers = map(int, np.floor(np.linspace(1, max_ngrams, 8)))
```

Comment:
Changed triers to num_ngrams and added np.unique. I think that doing CV for each added ngram sequentially would be too expensive, as on the test cases we were of order 100 extra ngrams and that would mean running cross validation a lot of times.